### ISSUE TITLE
willDisplay not called

### ISSUE THREAD

**Issue by Nicholmon** [patchthecode/JTAppleCalendar#1041]
IMPORTANT - BEFORE CREATING YOUR ISSUE PLEASE READ THE FOLLOWING:

1. PLEASE STATE THE VERSION NUMBER you are using.
7.1.7
2. Are you using the latest version of JTAppleCalendar? Latest version is currently 7.1.6
Yes. Latest downloadable is 7.1.7
3. PROVIDE ENOUGH INFORMATION SO I CAN RECREATE THE PROBLEM.
Adding a breakpoint in the `willDisplay` does not stop. I tried the same with the example app for and it does not get called either.
I am running it with Xcode 10.2

**Comment by patchthecode** [comment]
I am not sure what you are trying to do with that function, but according #553, it should have the same code as cellForItemAtDate.

**Comment by patchthecode** [comment]
It probably does not behave the way you think it does.
What are you trying to do?

**Comment by Nicholmon** [comment]
When I first populate the calendar the cells are somehow not configured according to their state but, once paginating across months in order to get the cells dequeued and refreshed, every cell has the correct UI.
My idea is that the automatic call of `willDisplay` will bind the cell once again showing the desired UI configuration. But this method is never called, not even while paginating.

Here are my 2 functions `cellForItemAt` and `willDisplay`:
```
func calendar(_ calendar: JTAppleCalendarView, cellForItemAt date: Date, cellState: CellState, indexPath: IndexPath) -> JTAppleCell {
    let cell = calendar.dequeueReusableJTAppleCell(withReuseIdentifier: CalendarDayCollectionViewCell.cellIdentifier, for: indexPath)
    if let cell = cell as? CalendarDayCollectionViewCell {
        cell.bind(cellState: cellState, selectable: dateInRange(date: date))
    }
    return cell
}
    
func calendar(_ calendar: JTAppleCalendarView, willDisplay cell: JTAppleCell, forItemAt date: Date, cellState: CellState, indexPath: IndexPath) {
    if let cell = cell as? CalendarDayCollectionViewCell {
        cell.bind(cellState: cellState, selectable: dateInRange(date: date))
    }
}
```

**Comment by Nicholmon** [comment]
> It probably does not behave the way you think it does.
> What are you trying to do?

I tried to place a breakpoint within the same method even in the example iOS app provided in the repo but even in that app (with no modification from me) it seems to have no effect; execution never hits the breakpoint

**Comment by patchthecode** [comment]
> When I first populate the calendar the cells are somehow not configured according to their state but

Then this is the problem.
We need to figure out why this is happening and solve it.

Using willDisplay will not work, because it was not coded that way. I will only be called in special situations as explained in #553 

I know your project might be huge. Therefore is it possible to recreate a empty calendar project which recreates the bad initial state? You dont have to design it or any thing. Just a simple calendar that loads with the bad state so that i can see why this is happening.

You can drop the zip fle in this chat box

**Comment by Nicholmon** [comment]
I will provide you with an example soon.
Could the issue be related to the fact that I am registering and loading cells from a xib and not creating them in the storyboard within the "calendar/collectionView"?
Because among the methods you expose you can register a nib for a reusableView only and not a cell..

**Comment by patchthecode** [comment]
No.
This lib works with xibs, storyboard and code.

So if there is an issue, it would be a bug that has to be fixed.

**Comment by Nicholmon** [comment]
I created a short video of what happens:
Data used to populate the calendar is fetched asynchronously.
Notice how at first loading, all dates in April have the same configuration. (They should change from 1 to 19 of April: see end of the video to see the expected UI configuration).
Bordered date is today.
When I paginate to next mont, dates have correct configuration. (For May).
When I paginate back, some dates have update configuration (correctly) but not all required.
If I perform a further pagination action to 2 months in the future (May and June) dates in these months appear correct; when I then come back to April again, cells are in the desired UI configuration.

[calendar.zip](https://github.com/patchthecode/JTAppleCalendar/files/3098125/calendar.zip)

I tried to preview the cell that is supposed to be returned by `cellForItemAt` and it shows the desired UI I would like to be displayed; unfortunately this is not how the cell is then rendered (as you can see in the video).

Here is the image of the 1st of April cell: at first loading it appears with wrong configuration but after first pagination already, it is displayed correctly.

From debug preview:
<img width="117" alt="Screenshot 2019-04-19 at 14 43 07" src="https://user-images.githubusercontent.com/27209147/56424720-c2e48380-62b1-11e9-8d5b-959478141794.png"> 
⬇️ What is first rendered:
<img width="56" alt="Screenshot 2019-04-19 at 14 45 57" src="https://user-images.githubusercontent.com/27209147/56424775-f2938b80-62b1-11e9-8260-4cabb962484a.png"> 
⬇️After first pagination and back:
<img width="47" alt="Screenshot 2019-04-19 at 14 46 08" src="https://user-images.githubusercontent.com/27209147/56424786-f9ba9980-62b1-11e9-8a97-36a779f6a8e1.png">

**Comment by patchthecode** [comment]
Can you try master branch and see if this issue is occurring? 

instructions [here](https://github.com/patchthecode/JTAppleCalendar/wiki/Common-Questions#common-questions-can-now-be-found-on-this-link)

It is important that you clear the build cache. Because sometimes even after pulling the new pod, it still has the old cache files. Sometimes i actually have to physically delete the `pod` folder and the `pod.lock` file together with the `build` folder generated by XCode. 

Let me know if the issue occurs on master branch code.

**Comment by Nicholmon** [comment]
I tried with this commit: a131d7fd2d447a81854194e4d0198519f7ea526e and the issue is still present

**Comment by Nicholmon** [comment]
@patchthecode Looks like I found a fix for it. At the beginning I was setting the delegate and data source of the calendar programmatically, not via storyboard, and probably this was creating the issue. I was doing that because the delegate configuration was called before the data was fetched and the configuration resulted then to be incorrect.
I refactored the view and population logic of the calendar to display it only after the completion of data fetching (as shown in the video) but I then did not switch back to the assignment of delegate and data source via storyboard.

The issue looks solved even with version 7.1.7, not only with latest `master`.

I will close this issue as it looks solved.
Thanks for the support!
